ReadWriteOncePod access mode alpha feature blog

chrishenzie opened the pull request:
<!-- 🛈

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/start/#improve-existing-content

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

This PR introduces the feature blog for the ReadWriteOncePod alpha.

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,104 @@
+In the world of Kubernetes, access modes are the way we model how storage is consumed.
+These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).
+
+<!-- TODO: Is there a way to strongly highlight the accessModes section inside this block? -->
```
We usually use YAML comments for that kind of highlighting.

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,104 @@
+
+## How do I use it?
+
+The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22. As a first step we need to enable the ReadWriteOncePod feature gate for `kube-apiserver`, `kube-scheduler`, and `kubelet`. The feature gate can be enabled or disabled as follows (as described in more detail [here]).
```
We'd typically write:
```suggestion
The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22. As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`. You can enable (or disable) the feature by setting command line arguments:
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,104 @@
+
+## As a storage vendor, how do I add support for this access mode to my CSI driver?
+
+### New Controller and Node Service Capabilities
```
```suggestion
### New controller and Node Service capabilities
```
Is that capitalization right? We usually put headings in sentence case.

chrishenzie replied:
"Node Service" here refers to the [container storage interface](https://github.com/container-storage-interface/spec/blob/master/spec.md#rpc-interface), and not the Kubernetes primitives; removing the capitalization

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,104 @@
+
+### How is this different than the ReadWriteOnce access mode?
+
+The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to write to the same volume.
```
Is this bad (potentially yes!) - let's explain the benefit. Maybe [fencing](https://en.wikipedia.org/wiki/Fencing_(computing)) a legacy app?

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,104 @@
+
+If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:
+
+```shell
````
This isn't a shell script. Best not to set any syntax highlighting.

sftim reviewed (commented):
Some early feedback

chrishenzie commented:
/assign @msau42 
/assign @alculquicondor

alculquicondor reviewed (commented):
I think blog post PRs go against the master branch.

chrishenzie commented:
@alculquicondor Thanks for the heads up. Changing the base branch in the PR produced all the commit spam above, sorry for any confusion.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+
+Could this cause problems?
+Potentially yes!
+If you are running workloads that access sensitive data, other pods on that node can access it (if updated to mount the volume).
```
I would not put security as the reason for this feature. We have other controls (PSP, PV being cluster scoped) to prevent pods from other namespaces to access the volume. Using RWOP for security is timing sensitive too. If the other workload gets scheduled first it will get access to the volume.

The primary reason I see is to provide data safety guarantees to applications that can only have at-most one writer.

chrishenzie replied:
Thank you for the clarification!

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+You don't have to worry about neighboring pods potentially accessing your volumes.
+Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.
+
+## How do I use it?
```
Note that this is only supported for CSI plugins.

As an aside, what happens if you enable this on an intree plugin?

chrishenzie replied:
Done. If you enable this on an intree plugin it should work, we just have to add ReadWriteOncePod to [the list of supported access modes](https://github.com/kubernetes/kubernetes/blob/v1.22.0-rc.0/pkg/volume/gcepd/gce_pd.go#L113-L118).

msau42 replied:
What was the reason why we decided not to support it for in-tree volumes if nothing stops a user from using it and it works?

chrishenzie replied:
See https://github.com/kubernetes/website/pull/29059#issuecomment-889289648

chrishenzie replied:
If we wanted to open this up to in-tree drivers we could do this for beta graduation.

chrishenzie replied:
Synced with @Jiawei0227 , I did not know the API fields for in-tree drivers were not being removed. If customers want to rely on the legacy API fields + CSI migration and we feel it will help target more users we can add support here

msau42 replied:
Yeah I think given that the enforcement is all done in Kubernetes, the plugin is not involved, and we don't error/reject if someone uses it with an in-tree type, it is better to support it for in-tree types as well so it's consistent across environments.

Jiawei0227 replied:
If the implementation of ReadWriteOncePod is done in kubelet or KCM not in each intree plugins code, I think it is fine to support it for intree as well. 

What I was thinking is if we need to edit intree plugin codes that will be not worth it as most of the PersistentVolume plugin types will be either deprecated or removed.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+
+If you have existing PersistentVolumes, they can be migrated to use ReadWriteOncePod.
+
+As a prerequisite, you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim:
```
Before you delete the PVC, you must change the PV.spec.reclaimPolicy to Retain, otherwise, the disk will get deleted.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+  name: cat-pictures-pv
+spec:
+  accessModes:
+    - ReadWriteOnce    # Delete existing access modes like ReadWriteOnce.
```
Maybe RWO should be commented out to facilitate someone copy-pasting the example.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+apiVersion: v1
+metadata:
+  name: cat-pictures-pv
+spec:
```
You should also set the spec.ClaimRef to the name/namespace of the PVC. That reserves the PV for that PVC and prevents the PV from being claimed by another PVC.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,269 @@
+Next you need to replace the PersistentVolume's access modes with ReadWriteOncePod:
+
+```yaml
+kind: PersistentVolume
````
You will probably want to set PVC.storageClassName to "", otherwise it will use the default storageclass.

Or set the PV.StorageClassName to the same storageClassName that the original PVC was.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+
+### Update your CSI sidecars (optional)
+
+When deplyoing your CSI drivers, you may update your CSI sidecars (mainly the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner), [external-attacher](https://github.com/kubernetes-csi/external-attacher), and [external-resizer](https://github.com/kubernetes-csi/external-resizer)) to versions that depend on CSI spec v1.5.0+ and the Kubernetes v1.22 API.
```
typo: deploying

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,269 @@
+* James DeFelice (jdef)
+* Humble Devassy Chirammal (humblec)
+* Aldo Culquicondor (alculquicondor)
+* ahg-g
```
Abdullah Gharaibeh, from https://github.com/kubernetes/community/tree/master/sig-scheduling

alculquicondor commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,301 @@
+--feature-gates="...,ReadWriteOncePod=true"
+```
+
+[here]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
````
is this line a left over?

chrishenzie replied:
Removed, thank you!

alculquicondor reviewed (commented):
Looks good for the scheduling bits

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,299 @@
+- ReadWriteMany -- the volume can be mounted as read-write by many nodes
+
+These access modes are enforced by Kubernetes components like the `kube-controller-manager` and `kubelet` to ensure only certain pods are allowed to access a given PersistentVolume.
+This allows users to follow the principle of least privilege to mitigate the impact of bad actors.
```
I would remove this line because AccessModes is not really a security feature.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,299 @@
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: ""
```
I think you'll want to have 2 different PVC examples: one for dynamic provisioning, and one for pre-existing volumes. 

For dynamic provisioning, the storageClassName should be empty (to use the default storageclass) or filled in.

chrishenzie replied:
Removed the `storageClassName` field.

This dynamic provisioning example is different (at least the name is), were there other changes I should make to differentiate?

chrishenzie replied:
Renamed this to `single-writer-only` (from `top-secret`) to clarify the intended use case for the access mode

chrishenzie commented:
@msau42 The reason I decided not to support it was because of the planned deprecation of in-tree drivers. Since they are going away, this is further incentive for customers to migrate to using CSI.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,297 @@
+As a first step, you need to edit your PersistentVolume's `spec.persistentVolumeReclaimPolicy` and set it to `Retain`.
+This ensures your PersistentVolume will not be deleted when we delete the corresponding PersistentVolumeClaim.
+
+```yaml
````
Is there an edit/patch command that can be used instead that can clearly show just the reclaim policy patch?

chrishenzie replied:
Done

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,297 @@
+
+#### Migrating existing PersistentVolumes
+
+If you have existing PersistentVolumes, they can be migrated to use ReadWriteOncePod.
```
Maybe also add something like "In this example, there is already a "cat-pictures-pvc" that is bound to a "cat-pictures-pv" PV."

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,297 @@
+kubectl delete pvc cat-pictures-pvc
+```
+
+After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:
````
Same thing, can we have a patch command that only changes the accessMode?

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,267 @@
+  storageClassName: ""
+```
+
+In Kubernetes, we currently have three access modes for PVs and PVCs:
````
```suggestion
Before v1.22, Kubernetes offered three access modes for PVs and PVCs:
```

sftim replied:
As this article will be published when **v1.22** is the current release, we really ought to fix the tense of the verb / sense of the text here. What's drafted here would be factually inaccurate at the time it is published, and might also mislead some readers until they read on.

chrishenzie replied:
Oops, I included this change but hadn't pushed it. Updated to use the suggested text

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+In Kubernetes, we currently have three access modes for PVs and PVCs:
+
+- ReadWriteOnce -- the volume can be mounted as read-write by a single node
+- ReadOnlyMany -- the volume can be mounted read-only by many nodes
+- ReadWriteMany -- the volume can be mounted as read-write by many nodes
```
nit: I would use an en dash here (`&ndash;`), rather than `--` (check the [preview](https://deploy-preview-29059--kubernetes-io-main-staging.netlify.app/blog/2021/07/16/read-write-once-pod-access-mode-alpha/) to see how these render)

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+## What is this new access mode and how does it work?
+
+In Kubernetes v1.22, a fourth access mode for PVs and PVCs is introduced for CSI volumes:
```
```suggestion
Kubernetes v1.22 introduced a fourth access mode for PVs and PVCs, that you can use for CSI volumes:
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+In Kubernetes v1.22, a fourth access mode for PVs and PVCs is introduced for CSI volumes:
+
+- ReadWriteOncePod -- the volume can be mounted as read-write by a single pod
```
nit: I would use an en dash here (`&ndash;`) too.

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+- ReadWriteOncePod -- the volume can be mounted as read-write by a single pod
+
+When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.
```
I couldn't work out from the existing wording whether ReadWriteOncePod blocks another Pod having just read access to that nonvolatile storage.
```suggestion
If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that new pod is the only pod across your whole the cluster that can read that PVC or write to it.
```

chrishenzie replied:
Thank you for the suggestion. The tricky thing about this wording is "new", it implies a new pod will usurp control over the PVC if it's already in-use by another pod. I kept the sentence but removed the word "new", does it make sense without it?

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.
+
+If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:
```
```suggestion
If you create another pod that references the same PVC then, no matter what access mode
you suggest for the new pod, the ReadWriteOncePod access mode on the existing pod
means that the new pod fails to start. For example:
```

chrishenzie replied:
These access modes are not on pods, but on the PVC. I provided an explanation for why it won't start. How does this look?

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.
+This could potentially be a major problem for some applications, especially if they require at most one writer for data safety guarantees.
+
+With ReadWriteOncePod these issues go away.
+Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.
```
For technical writing, I try to avoid “just” in this sense - and especially when there _are_ extra steps, such as enabling feature gates across multiple components.

```suggestion
With ReadWriteOncePod these issues go away:
set the access mode on your PVC, and Kubernetes guarantees that only a single pod has access.
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,267 @@
+kubectl edit pv cat-pictures-pv
+```
+
+[Please see this guide](https://kubernetes.io/docs/tasks/configure-pod-container/configure-persistent-volume-storage/) for more details on working with PersistentVolumes and PersistentVolumeClaims.
````
```suggestion
You can read [Configure a Pod to Use a PersistentVolume for Storage](/docs/tasks/configure-pod-container/configure-persistent-volume-storage/) for more details on working with PersistentVolumes and PersistentVolumeClaims.
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+## What volume plugins support this?
+
+The only volume plugins that support this are CSI drivers.
+We do not plan to support this for in-tree plugins because they are being deprecated as part of [CSI migration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/#what-is-the-timeline-status).
```
Readers might not know if “we” refers to Google, the Kubernetes project, or SIG Storage. Can you clarify?

chrishenzie replied:
Clarified to SIG Storage

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+They are [`SINGLE_NODE_SINGLE_WRITER` and `SINGLE_NODE_MULTI_WRITER`](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L437-L447).
+In order to communicate to sidecars (like the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner)) that your driver understands and accepts these two new CSI access modes, your driver will also need to advertise the `SINGLE_NODE_MULTI_WRITER` capability for the [controller service](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L1073-L1081) an ...
+
+[Please see the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes-volume-capabilities) for more context on the motivation for these access modes and capability bits.
```
```suggestion
If you'd like to read up on the motivation for these access modes and capability bits, you can also read
the [CSI Specification Changes, Volume Capabilities](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes-volume-capabilities)
section of KEP-2485 (ReadWriteOncePod PersistentVolume Access Mode).
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+[Please see the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes-volume-capabilities) for more context on the motivation for these access modes and capability bits.
+
+### Update your CSI spec
```
```suggestion
### Update your CSI driver to use the new interface
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+## What’s next?
+
+As part of the beta graduation for this feature, support will be added for pod preemption in the scheduler.
```
- I'd write “we plan to” for statements like this (exception: SIG Architecture have signed off on a commitment that it will happen this way)
- the scheduler already supports preemption; I've reworded to clarify what would be added
```suggestion
As part of the beta graduation for this feature, we plan to update the Kubenetes scheduler to support pod preemption in relation to ReadWriteOncePod storage.
```

(also see separate comment about making sure the reader is clear who “we” refers to).

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+**Author:** Chris Henzie (Google)
+
+Kubernetes 1.22 introduces a new ReadWriteOncePod access mode for PersistentVolumes and PersistentVolumeClaims that allows users to restrict volume access to a single pod in the cluster.
```
Optionally: link PersistentVolumes and PersistentVolumeClaims to https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes and https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims (respectively)

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+
+## What are access modes and why are they important?
+
+When using storage, there are different ways we can model how it is consumed.
```
I'd write:
```suggestion
When using storage, there are different ways to model how that storage is consumed.
```

How does that sound?

chrishenzie replied:
Sounds good to me!

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,267 @@
+In other cases maybe everyone is allowed to read data but not write it.
+For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.
+
+In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way we model how storage is consumed.
```
```suggestion
In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way you can define how durable storage is consumed.
```
?

chrishenzie replied:
Thanks!

sftim reviewed (commented):
Thanks for this PR @chrishenzie 

Here's some feedback; I hope it's useful. Please do reply if you have any queries and especially if you disagree with any of the points.

chrishenzie commented:
@sftim I updated the links to remove `https://kubernetes.io` to make them relative to the host, is that preferred?

sftim commented:
> updated the links to remove https://kubernetes.io to make them relative to the host, is that preferred?

It is preferred but for blog articles not essential to get right.

> should all references to 1.22 (or other K8s versions) have the "v" prefix?

I don't think I could write a style guide entry to explain it but I do get a sense of when the project typically would or would not use the prefix.
It's OK to use your own preference there, this is _your_ article. So long as things are clear.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,268 @@
+Lastly you may edit your PersistentVolume's `spec.persistentVolumeReclaimPolicy` and set to it back to `Delete` if you previously changed it.
+
+```
+kubectl edit pv cat-pictures-pv
````
use the patch command again

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,268 @@
+## What volume plugins support this?
+
+The only volume plugins that support this are CSI drivers.
+SIG Storage does not plan to support this for in-tree plugins because they are being deprecated as part of [CSI migration](/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/#what-is-the-timeline-status).
```
Do we want to modify this to say something like "we may consider support for beta"

chrishenzie replied:
Added:
"Support may be considered for beta for users that prefer to use the legacy in-tree volume APIs with CSI migration enabled."

sftim commented:
/sig storage

PI-Victor commented:
folks, this needs to be targeted at the `main` branch, not at `dev-1.22`.

PI-Victor commented:
since this will be merged post-release, i'm removing this from the 1.22 milestone, no worries, this does not affect the publishing of the blog or anything else.
/milestone clear

sftim commented:
/retest

sftim commented:
@kubernetes/sig-storage-misc - it'd be good to get a confirmation that this article looks technically accurate.

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
````diff
@@ -0,0 +1,269 @@
+After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:
+
+```shell
+kubectl patch pv cat-pictures-pv -p '{"spec":{"accessModes":["ReadWriteOncePod"]}}'
````
Oh you may also need to remove the pv.spec.ClaimRef.UID field in order to rebind to the new PVC. You may need to test out this sequence to make sure it works

chrishenzie replied:
Thank you for pointing this out, I encountered this issue when running through the steps

chrishenzie commented:
I've updated the instructions for migrating an existing PV to use ReadWriteOncePod. PTAL when you have a moment. I've ran through them and they work.

sftim commented:
@kubernetes/sig-storage-misc (maybe @msau42) - how's this looking now?

msau42 commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,280 @@
+
+Next you need to modify your PersistentVolumeClaim to set ReadWriteOncePod as the only access mode.
+You should also set your PersistentVolumeClaim's `spec.volumeName` to the name of your PersistentVolume.
+Doing this guarantees your PersistentVolumeClaim will bind to this specific PersistentVolume.
```
It's actually the opposite (setting PV.spec.claimref) that guarantees the binding. If you only set the PVC.spec.volumeName, you could run into an issue where multiple PVCs set the same PV.

msau42 reviewed (commented):
Just a nit, otherwise lgtm

msau42 commented:
/lgtm

chrishenzie commented:
I added one small section after the feature-gate to specify the minimum CSI sidecar versions required for this feature to work.

tengqm commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+
+- ReadWriteOncePod &ndash; the volume can be mounted as read-write by a single pod
+
+If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that pod is the only pod across your whole the cluster that can read that PVC or write to it.
```
```suggestion
If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that pod is the only pod across your whole cluster that can read that PVC or write to it.
```

chrishenzie replied:
Oops, fixed

tengqm commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+
+The ReadWriteOncePod access mode is in alpha for Kubernetes v1.22 and is only supported for CSI volumes.
+As a first step you need to enable the ReadWriteOncePod [feature gate](/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.
+You can enable (or disable) the feature by setting command line arguments:
```
```suggestion
You can enable the feature by setting command line arguments:
```

msau42 commented:
/lgtm

geoffcline commented:
I read through this. I think it's fine to still publish this. This may have been intended as a launch blog, but the content is still relevant.

For example, the statement "Kubernetes v1.22 introduced a fourth access mode for PVs and PVC" is still valid and relevant.

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+
+**Author:** Chris Henzie (Google)
+
+Kubernetes v1.22 introduces a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) that allows users to restrict volume access to a single pod in the cluster.
```
```suggestion
Last month's release of Kubernetes v1.22 introduced a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims). With this alpha feature, Kubernetes allows you to restrict volume access to a single pod in the cluster.
```
_or_
```suggestion
Last month's release of Kubernetes v1.22 introduced a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims). With this alpha feature, Kubernetes allows you to define volumes that only a single pod can access at a time.
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+  resources:
+    requests:
+      storage: 1Gi
+  storageClassName: ""
```
Is that empty value right? Could we omit it?

chrishenzie replied:
Yes we can, it will default to the customer's default storage class (we do the same thing with the example on like 93).

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+The CSI spec outlines expected behavior for the `NodePublishVolume` RPC when called more than once for the same volume but with different arguments (like the target path).
+Please refer to [the second table in the NodePublishVolume section of the CSI spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodepublishvolume) for more details on expected behavior when implementing in your driver.
+
+### Update your CSI sidecars (optional)
```
only a nit, totally fine to leave as-is:
```suggestion
### Update your CSI sidecars (optional) {#update-csi-sidecars}
```

sftim commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,285 @@
+
+Special thanks to the following people for their insightful reviews and design considerations:
+
+* Michelle Au (msau42)
+* Ben Swartzlander (bswartz)
+* Jan Šafránek (jsafrane)
+* Hemant Kumar (gnufied)
+* Xing Yang (xing-yang)
+* Saad Ali (saad-ali)
+* Jing Xu (jingxu97)
+* Tim Hockin (thockin)
+* Jordan Liggitt (liggitt)
+* Deep Debroy (ddebroy)
+* James DeFelice (jdef)
+* Humble Devassy Chirammal (humblec)
+* Aldo Culquicondor (alculquicondor)
+* Abdullah Gharaibeh (ahg-g)
```
nit: I might consider sorting that list alphabetically.

sftim reviewed (commented):
/hold

We missed the publication date. However, with only small changes this would definitely be good to publish.

sftim commented:
Let's pick a new date. How does the 13th of September sound?

chrishenzie commented:
@sftim @msau42 

A summary of the latest changes:

- Changed intro sentence to specify "last month's release"
- Removed storage class name in first example
- Make CSI sidecar update a required step
  - Updating the driver is optional but updating sidecars is required if they want to use the feature
  - This is specified at the top of the article and now additionally at the bottom in the instructions for adding support in your CSI driver
- Alphabetized reviewer names

Would you prefer the date for this article be updated as well?

msau42 commented:
/lgtm

geoffcline commented on content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md:
```diff
@@ -0,0 +1,287 @@
+---
+layout: blog
+title: "Introducing Single Pod Access Mode for PersistentVolumes"
+date: 2021-07-16
```
This is the date that needs to be updated.

I'd use the date Tim suggested.

geoffcline commented:
Thanks for updating the date!

/lgtm

sftim commented:
/hold cancel
/approve

Thanks folks - @chrishenzie especially!
